scripts/32-rust-is-written-in-rust.md

```diff
@@ -0,0 +1,790 @@
+
+It's a fantastic product made by some nice people, and I'm delighted to say they are hiring.
+
+---
+
+<!-- slide bg="rgb(37, 34, 43)" -->
+
+## Founding Software Engineer
+
+#### (Rust, WASM)
+
+<br/>
+More info:
+
+### <https://careers.quadratichq.com>
+
+notes:
+
+Quadratic are looking for:
+- Great Rust engineers who have experience at a Startup and leading software architecture and implementation.
+
+---
+
+<!-- slide bg="rgb(37, 34, 43)" -->
+
+![[Quadratic Logo.png]]
+
+The infinite canvas spreadsheet with code
+
+#### https://QuadraticHQ.com
+
+#### <https://careers.quadratichq.com>
+
+notes:
+
+-   Apply today, and
+-   Head to QuadraticHQ.com to try it out.
+
+My thanks to quadratic for their support of this channel.
+
+Back to our archaeological dig.
+
+---
+
+## Options
+
+```rust[]
+pub enum Option<T> {
+    None,
+    Some(T),
+}
+```
+
+notes:
+
+Rust's option type is not an opaque language feature, it's built in Rust's sum type: Enums.
+
+Enums are a core component of Rust's rich type system.
+
+You or I could build our own, should we wish to.
+
+A fantastic, clear explanation of sum types is in this pinned video by Logan Smith by the way
+https://www.youtube.com/watch?v=s5S2Ed5T-dc
+
+And for practical examples of using enums, you can watch my video "rust data modelling without classes"
+
+---
+
+```rust
+fn div(dividend: i32, divisor: i32) -> Option<i32> {
+    if divisor == 0 {
+        None
+    } else {
+        Some(dividend / divisor)
+    }
+}
+fn try_division(dividend: i32, divisor: i32) {
+    match div(dividend, divisor) {
+        None => println!("{dividend} / {divisor} fail"),
+        Some(quotient) => {
+            println!("{dividend}/{divisor}={quotient}");
+        },
+}}
+```
+
+notes:
+
+Rust does not have the concept of Null, instead using options ubiquitously throughout the language.
+
+Nearly all other languages, old standards such as javascript, python, and java, and even recently-designed ones like Go and Kotlin, began with nulls, the billion-dollar mistake, only to realise their error and attempt a retro-fit of options.
+
+This never works, and you can feel it in the ecosystem and even standard libraries of these languages.
+
+The Rust universe, just like the mathematical universe doesn't have the human concept of nulls, an exception to the type system.
+
+---
+
+## [std::result::Result](https://doc.rust-lang.org/std/result/enum.Result.html)
+
+```rust
+pub enum Result<T, E> {
+    Ok(T),
+    Err(E),
+}
+```
+
+notes:
+
+If you have sum types, rust enums, you can build enormous parts of your standard library with this simple structure instead of building it in to the plumbing of the compiler.
+
+You can use the rules of your universe to make what you want.
+
+Here is the result enum, rust's answer to error passing.
+
+---
+
+```rust[]
+let mut file = match File::open(&path) {
+    Err(why) => panic!("can't open {path.display()}"),
```

Comment:
This is unfortunately not valid rust. You can't have a function call or a dereference in a format argument.